walnuts_cpp/include/walnuts/nuts.hpp

```diff
@@ -122,14 +130,29 @@ Span<S> combine(Random<S, Generator> &rng, Span<S> &&span_old,
   S update_logprob = span_new.logp_ - log_denominator;
   bool update = log(rng.uniform_real_01()) < update_logprob;
   auto &selected = update ? span_new.theta_select_ : span_old.theta_select_;
-  if constexpr (D == Direction::Forward) {
-    uturn_flag = uturn(span_old, span_new, inv_mass);
-    return Span<S>(std::move(span_old), std::move(span_new),
-                   std::move(selected), logp_total);
-  } else { // Direction::Backward
-    uturn_flag = uturn(span_new, span_old, inv_mass);
-    return Span<S>(std::move(span_new), std::move(span_old),
-                   std::move(selected), logp_total);
+
+  bool did_uturn = uturn<D>(span_old, span_new, inv_mass);
+
+  if constexpr (R == CombineResult::ThrowAway) {
+    if (did_uturn) {
+      return std::nullopt;
+    }
+  }
+
+  Span<S> next = [&]() {
```

Comment:
Now this is really getting complicated.  This one I can't figure out in terms of what `std::move(next)` does when `next` is a closure.  From first principles, it should send an rvalue reference to `make_pair` or `make_optional` so that they can steal its memory.  

I'd personally rather write two simple functions than one complicated function like this, and then just delegate to any shared behavior.

Reply:
This is actually not a closure, it's an inline lambda that is immediately evaluated. This is one possible workaround for the fact that there is no `if constexpr` at the expression/ternary level. But if it's too cute, I'm happy to look into other alternatives

Reply:
I can see that it's getting evaluated right away.  They mention this pattern in the link I included.  Isn't the nested lambda technically a closure because it captures environmental variables such as `span_old` with `[&]`?

Reply:
Oh, the construct itself is a closure, it's just that you said

> This one I can't figure out in terms of what `std::move(next)` does when `next` is a closure

But `next` is not a closure, it's the result of evaluating one, which makes it just a normal `Span`

Reply:
My bad.  I should've seen that before as it has an explicit type declaration!